File: go/internal/dry/copied_blocks.go
package dry

import (
	"go/scanner"
	"go/token"
	"sort"
	"strconv"
	"strings"
)

type tokenEntry struct {
	tag  string
	file string
	line int
}

type tokenSequence struct {
	file   sourceFile
	tokens []tokenEntry
}

type tokenOccurrence struct {
	sequence int
	index    int
}

func findCopiedBlocks(files []sourceFile, options Options) ([]Finding, error) {
	sequences, err := tokenSequences(files)
	if err != nil {
		return nil, err
	}
	matches := copiedBlockMatches(sequences, options.CopiedBlockTokens)
	sortCopiedMatches(matches)
	return copiedBlockFindings(matches), nil
}

func copiedBlockMatches(sequences []tokenSequence, windowSize int) []Finding {
	windows := map[string][]tokenOccurrence{}
	for sequenceIndex, sequence := range sequences {
		for i := 0; i+windowSize <= len(sequence.tokens); i++ {
			key := tokenWindowKey(sequence.tokens[i : i+windowSize])
			windows[key] = append(windows[key], tokenOccurrence{sequence: sequenceIndex, index: i})
		}
	}

	seen := map[string]bool{}
	var matches []Finding
	for _, occurrences := range windows {
		matches = append(matches, matchesForOccurrences(sequences, occurrences, windowSize, seen)...)
	}
	return matches
}

func matchesForOccurrences(
	sequences []tokenSequence,
	occurrences []tokenOccurrence,
	windowSize int,
	seen map[string]bool,
) []Finding {
	var matches []Finding
	for i := 0; i < len(occurrences); i++ {
		for j := i + 1; j < len(occurrences); j++ {
			match, ok := copiedBlockMatch(sequences, occurrences[i], occurrences[j], windowSize, seen)
			if ok {
				matches = append(matches, match)
			}
		}
	}
	return matches
}

func copiedBlockMatch(
	sequences []tokenSequence,
	left tokenOccurrence,
	right tokenOccurrence,
	windowSize int,
	seen map[string]bool,
) (Finding, bool) {
	if occurrenceLess(right, left, sequences) {
		left, right = right, left
	}
	match := expandTokenMatch(sequences, left, right, windowSize)
	if match.Tokens < windowSize {
		return Finding{}, false
	}
	key := copiedFindingKey(match)
	if seen[key] {
		return Finding{}, false
	}
	seen[key] = true
	return match, true
}

func sortCopiedMatches(matches []Finding) {
	sort.Slice(matches, func(i, j int) bool {
		left := matches[i]
		right := matches[j]
		if left.Tokens != right.Tokens {
			return left.Tokens > right.Tokens
		}
		return rangeKey(left.Left, left.Right) < rangeKey(right.Left, right.Right)
	})
}

func copiedBlockFindings(matches []Finding) []Finding {
	var findings []Finding
	for _, match := range matches {
		if overlapsCopiedFinding(match, findings) {
			continue
		}
		findings = append(findings, match)
	}
	sort.Slice(findings, func(i, j int) bool {
		return rangeKey(findings[i].Left, findings[i].Right) < rangeKey(findings[j].Left, findings[j].Right)
	})
	return findings
}

func tokenSequences(files []sourceFile) ([]tokenSequence, error) {
	sequences := make([]tokenSequence, 0, len(files))
	for _, file := range files {
		fileSet := token.NewFileSet()
		scanFile := fileSet.AddFile(file.Path, fileSet.Base(), len(file.Content))
		var lexer scanner.Scanner
		lexer.Init(scanFile, file.Content, nil, 0)
		var tokens []tokenEntry
		for {
			pos, tok, lit := lexer.Scan()
			if tok == token.EOF {
				break
			}
			tag, ok := normalizedToken(tok, lit)
			if !ok {
				continue
			}
			position := fileSet.Position(pos)
			tokens = append(tokens, tokenEntry{tag: tag, file: file.Path, line: position.Line})
		}
		sequences = append(sequences, tokenSequence{file: file, tokens: tokens})
	}
	return sequences, nil
}

func normalizedToken(tok token.Token, lit string) (string, bool) {
	switch tok {
	case token.COMMENT, token.SEMICOLON:
		return "", false
	case token.IDENT:
		return "ident/" + lit, true
	case token.INT, token.FLOAT, token.IMAG, token.CHAR, token.STRING:
		return "literal/" + tok.String() + "/" + lit, true
	default:
		return tok.String(), true
	}
}

func tokenWindowKey(tokens []tokenEntry) string {
	var builder strings.Builder
	for _, token := range tokens {
		builder.WriteString(token.tag)
		builder.WriteByte(0)
	}
	return builder.String()
}

func occurrenceLess(left, right tokenOccurrence, sequences []tokenSequence) bool {
	leftFile := sequences[left.sequence].file.Path
	rightFile := sequences[right.sequence].file.Path
	if leftFile != rightFile {
		return leftFile < rightFile
	}
	return left.index < right.index
}

func expandTokenMatch(sequences []tokenSequence, left, right tokenOccurrence, windowSize int) Finding {
	leftTokens := sequences[left.sequence].tokens
	rightTokens := sequences[right.sequence].tokens
	startLeft, startRight := expandTokenStart(leftTokens, rightTokens, left, right, windowSize)
	endLeft, endRight := expandTokenEnd(leftTokens, rightTokens, left, right, startLeft, startRight, windowSize)

	return Finding{
		Kind:   "copied-block",
		Left:   tokenRange(leftTokens[startLeft:endLeft]),
		Right:  tokenRange(rightTokens[startRight:endRight]),
		Tokens: endLeft - startLeft,
		Engine: "token-window",
	}
}

func expandTokenStart(
	leftTokens []tokenEntry,
	rightTokens []tokenEntry,
	left tokenOccurrence,
	right tokenOccurrence,
	windowSize int,
) (int, int) {
	startLeft := left.index
	startRight := right.index
	for startLeft > 0 && startRight > 0 && leftTokens[startLeft-1].tag == rightTokens[startRight-1].tag {
		if left.sequence == right.sequence && rangesOverlap(startLeft-1, left.index+windowSize-1, startRight-1, right.index+windowSize-1) {
			break
		}
		startLeft--
		startRight--
	}
	return startLeft, startRight
}

func expandTokenEnd(
	leftTokens []tokenEntry,
	rightTokens []tokenEntry,
	left tokenOccurrence,
	right tokenOccurrence,
	startLeft int,
	startRight int,
	windowSize int,
) (int, int) {
	endLeft := left.index + windowSize
	endRight := right.index + windowSize
	for endLeft < len(leftTokens) && endRight < len(rightTokens) && leftTokens[endLeft].tag == rightTokens[endRight].tag {
		if left.sequence == right.sequence && rangesOverlap(startLeft, endLeft, startRight, endRight) {
			break
		}
		endLeft++
		endRight++
	}
	return endLeft, endRight
}

func rangesOverlap(leftStart, leftEnd, rightStart, rightEnd int) bool {
	return max(leftStart, rightStart) <= min(leftEnd, rightEnd)
}

func tokenRange(tokens []tokenEntry) Range {
	return Range{
		Path:      tokens[0].file,
		StartLine: tokens[0].line,
		EndLine:   tokens[len(tokens)-1].line,
	}
}

func copiedFindingKey(finding Finding) string {
	return strings.Join([]string{
		finding.Left.Path,
		strconv.Itoa(finding.Left.StartLine),
		strconv.Itoa(finding.Left.EndLine),
		finding.Right.Path,
		strconv.Itoa(finding.Right.StartLine),
		strconv.Itoa(finding.Right.EndLine),
		strconv.Itoa(finding.Tokens),
	}, "|")
}

func overlapsCopiedFinding(candidate Finding, findings []Finding) bool {
	for _, existing := range findings {
		if existing.Kind != "copied-block" {
			continue
		}
		if sameRangePair(candidate.Left, candidate.Right, existing.Left, existing.Right) ||
			sameRangePair(candidate.Left, candidate.Right, existing.Right, existing.Left) {
			return true
		}
	}
	return false
}

func sameRangePair(left, right, existingLeft, existingRight Range) bool {
	return left.Path == existingLeft.Path &&
		right.Path == existingRight.Path &&
		rangesOverlapRange(left, existingLeft) &&
		rangesOverlapRange(right, existingRight)
}

File: typescript/src/dry/copied-blocks.ts
import { rangesOverlap } from "./report.js";
import { tokenSequences, type TokenEntry, type TokenSequence } from "./tokenize.js";
import type { DryFinding, SourceFile, SourceRange } from "./types.js";

export function findCopiedBlocks(
  files: readonly SourceFile[],
  windowSize: number
): readonly DryFinding[] {
  const matches = copiedBlockMatches(tokenSequences(files), windowSize);
  return collapseOverlapping(matches);
}

type Occurrence = {
  readonly sequence: number;
  readonly index: number;
};

function copiedBlockMatches(
  sequences: readonly TokenSequence[],
  windowSize: number
): readonly DryFinding[] {
  const windows = new Map<string, Occurrence[]>();
  sequences.forEach((sequence, sequenceIndex) => {
    for (let index = 0; index + windowSize <= sequence.tokens.length; index++) {
      const key = tokenWindowKey(sequence.tokens.slice(index, index + windowSize));
      windows.set(key, [...(windows.get(key) ?? []), { sequence: sequenceIndex, index }]);
    }
  });
  return [...windows.values()].flatMap((occurrences) =>
    matchesForOccurrences(sequences, occurrences, windowSize)
  );
}

function matchesForOccurrences(
  sequences: readonly TokenSequence[],
  occurrences: readonly Occurrence[],
  windowSize: number
): readonly DryFinding[] {
  const seen = new Set<string>();
  const matches: DryFinding[] = [];
  for (const pair of adjacentOccurrencePairs(
    compactOccurrences(sequences, occurrences, windowSize)
  )) {
    const match = copiedBlockMatch(sequences, pair.left, pair.right, windowSize);
    if (match !== undefined && !seen.has(findingKey(match))) {
      seen.add(findingKey(match));
      matches.push(match);
    }
  }
  return matches;
}

type OccurrencePair = {
  readonly left: Occurrence;
  readonly right: Occurrence;
};

function compactOccurrences(
  sequences: readonly TokenSequence[],
  occurrences: readonly Occurrence[],
  windowSize: number
): readonly Occurrence[] {
  const lastIndexBySequence = new Map<number, number>();
  const compacted: Occurrence[] = [];
  for (const occurrence of sortOccurrences(sequences, occurrences)) {
    const lastIndex = lastIndexBySequence.get(occurrence.sequence);
    if (lastIndex !== undefined && occurrence.index < lastIndex + windowSize) {
      continue;
    }
    compacted.push(occurrence);
    lastIndexBySequence.set(occurrence.sequence, occurrence.index);
  }
  return compacted;
}

function adjacentOccurrencePairs(occurrences: readonly Occurrence[]): readonly OccurrencePair[] {
  const pairs: OccurrencePair[] = [];
  for (let index = 0; index + 1 < occurrences.length; index++) {
    const left = occurrences[index];
    const right = occurrences[index + 1];
    if (left !== undefined && right !== undefined) {
      pairs.push({ left, right });
    }
  }
  return pairs;
}

function sortOccurrences(
  sequences: readonly TokenSequence[],
  occurrences: readonly Occurrence[]
): readonly Occurrence[] {
  return [...occurrences].sort((left, right) => compareOccurrences(left, right, sequences));
}

function copiedBlockMatch(
  sequences: readonly TokenSequence[],
  leftInput: Occurrence | undefined,
  rightInput: Occurrence | undefined,
  windowSize: number
): DryFinding | undefined {
  if (leftInput === undefined || rightInput === undefined) {
    return undefined;
  }
  const [left, right] = occurrenceLess(rightInput, leftInput, sequences)
    ? [rightInput, leftInput]
    : [leftInput, rightInput];
  if (
    left.sequence === right.sequence &&
    tokenRangesOverlap(
      left.index,
      left.index + windowSize - 1,
      right.index,
      right.index + windowSize - 1
    )
  ) {
    return undefined;
  }
  const match = expandTokenMatch(sequences, left, right, windowSize);
  return match.tokens >= windowSize ? match : undefined;
}

function expandTokenMatch(
  sequences: readonly TokenSequence[],
  left: Occurrence,
  right: Occurrence,
  windowSize: number
): DryFinding {
  const leftTokens = sequences[left.sequence]?.tokens ?? [];
  const rightTokens = sequences[right.sequence]?.tokens ?? [];
  const starts = expandStart(leftTokens, rightTokens, left, right, windowSize);
  const ends = expandEnd(leftTokens, rightTokens, left, right, starts, windowSize);
  return {
    kind: "copied-block",
    left: tokenRange(leftTokens.slice(starts.left, ends.left)),
    right: tokenRange(rightTokens.slice(starts.right, ends.right)),
    tokens: ends.left - starts.left,
    engine: "token-window"
  };
}

type Pair = {
  readonly left: number;
  readonly right: number;
};

function expandStart(
  leftTokens: readonly TokenEntry[],
  rightTokens: readonly TokenEntry[],
  left: Occurrence,
  right: Occurrence,
  windowSize: number
): Pair {
  let leftStart = left.index;
  let rightStart = right.index;
  while (
    leftStart > 0 &&
    rightStart > 0 &&
    leftTokens[leftStart - 1]?.tag === rightTokens[rightStart - 1]?.tag
  ) {
    if (
      left.sequence === right.sequence &&
      tokenRangesOverlap(
        leftStart - 1,
        left.index + windowSize - 1,
        rightStart - 1,
        right.index + windowSize - 1
      )
    ) {
      break;
    }
    leftStart--;
    rightStart--;
  }
  return { left: leftStart, right: rightStart };
}

function expandEnd(
  leftTokens: readonly TokenEntry[],
  rightTokens: readonly TokenEntry[],
  left: Occurrence,
  right: Occurrence,
  starts: Pair,
  windowSize: number
): Pair {
  let leftEnd = left.index + windowSize;
  let rightEnd = right.index + windowSize;
  while (
    leftEnd < leftTokens.length &&
    rightEnd < rightTokens.length &&
    leftTokens[leftEnd]?.tag === rightTokens[rightEnd]?.tag
  ) {
    if (
      left.sequence === right.sequence &&
      tokenRangesOverlap(starts.left, leftEnd, starts.right, rightEnd)
    ) {
      break;
    }
    leftEnd++;
    rightEnd++;
  }
  return { left: leftEnd, right: rightEnd };
}

function collapseOverlapping(matches: readonly DryFinding[]): readonly DryFinding[] {
  const sorted = [...matches].sort(
    (left, right) => right.tokens - left.tokens || findingKey(left).localeCompare(findingKey(right))
  );
  const kept: DryFinding[] = [];
  for (const match of sorted) {
    if (!kept.some((existing) => overlapPair(match, existing))) {
      kept.push(match);
    }
  }
  return kept.sort((left, right) => findingKey(left).localeCompare(findingKey(right)));
}

function overlapPair(left: DryFinding, right: DryFinding): boolean {
  return (
    (rangesOverlap(left.left, right.left) && rangesOverlap(left.right, right.right)) ||
    (rangesOverlap(left.left, right.right) && rangesOverlap(left.right, right.left))
  );
}

function occurrenceLess(
  left: Occurrence,
  right: Occurrence,
  sequences: readonly TokenSequence[]
): boolean {
  return compareOccurrences(left, right, sequences) < 0;
}

function compareOccurrences(
  left: Occurrence,
  right: Occurrence,
  sequences: readonly TokenSequence[]
): number {
  const leftFile = sequences[left.sequence]?.file.path ?? "";
  const rightFile = sequences[right.sequence]?.file.path ?? "";
  const byFile = leftFile.localeCompare(rightFile);
  if (byFile !== 0) {
    return byFile;
  }
  return left.sequence === right.sequence
    ? left.index - right.index
    : left.sequence - right.sequence;
}

function tokenWindowKey(tokens: readonly TokenEntry[]): string {
  return tokens.map((token) => token.tag).join("\0");
}

function tokenRange(tokens: readonly TokenEntry[]): SourceRange {
  const first = tokens[0];
  const last = tokens[tokens.length - 1];
  if (first === undefined || last === undefined) {
    return { path: "", start_line: 0, end_line: 0 };
  }
  return { path: first.file, start_line: first.line, end_line: last.line };
}

function tokenRangesOverlap(
  leftStart: number,
  leftEnd: number,
  rightStart: number,
  rightEnd: number
): boolean {
  return Math.max(leftStart, rightStart) <= Math.min(leftEnd, rightEnd);
}

function findingKey(finding: DryFinding): string {
  return [
    finding.left.path,
    finding.left.start_line,
    finding.left.end_line,
    finding.right.path,
    finding.right.start_line,
    finding.right.end_line,
    finding.tokens
  ].join("|");
}
